Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -3585,27 +3643,561 @@ class BBQJIT {
+            size_t elementSize = elementType.as<PackedType>() == PackedType::I8 ? sizeof(uint8_t) : sizeof(uint16_t);
+            uint8_t bitShift = (sizeof(uint32_t) - elementSize) * 8;
+            Location resultLocation = allocate(result);
+
+            m_jit.lshift32(TrustedImm32(bitShift), resultLocation.asGPR());
+            m_jit.rshift32(TrustedImm32(bitShift), resultLocation.asGPR());
+            LOG_INSTRUCTION("ArrayGetS", typeIndex, arrayref, index, RESULT(result));
+            break;
+        }
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
+            return { };
+        }
+        return { };
+    }
+
+    PartialResult WARN_UNUSED_RETURN addArraySet(uint32_t typeIndex, ExpressionType arrayref, ExpressionType index, ExpressionType value)
+    {
+        if (arrayref.isConst()) {
+            ASSERT(arrayref.asI64() == JSValue::encode(jsNull()));
+            emitThrowException(ExceptionType::NullArraySet);
+            return { };
+        }
+
+        Location arrayLocation = loadIfNecessary(arrayref);
+        emitThrowOnNullReference(ExceptionType::NullArraySet, arrayLocation);
+
+        if (index.isConst()) {
+            m_jit.load32(MacroAssembler::Address(arrayLocation.asGPR(), JSWebAssemblyArray::offsetOfSize()), wasmScratchGPR);
+            throwExceptionIf(ExceptionType::OutOfBoundsArraySet,
+                m_jit.branch32(MacroAssembler::BelowOrEqual, wasmScratchGPR, TrustedImm32(index.asI32())));
+        } else {
+            Location indexLocation = loadIfNecessary(index);
+            throwExceptionIf(ExceptionType::OutOfBoundsArraySet,
+                m_jit.branch32(MacroAssembler::AboveOrEqual, indexLocation.asGPR(), MacroAssembler::Address(arrayLocation.asGPR(), JSWebAssemblyArray::offsetOfSize())));
+        }
+
+        emitArraySetUnchecked(typeIndex, arrayref, index, value);
+
+        LOG_INSTRUCTION("ArraySet", typeIndex, arrayref, index, value);
+        return { };
+    }
+
+    PartialResult WARN_UNUSED_RETURN addArrayLen(ExpressionType arrayref, ExpressionType& result)
+    {
+        if (arrayref.isConst()) {
+            ASSERT(arrayref.asI64() == JSValue::encode(jsNull()));
+            emitThrowException(ExceptionType::NullArrayLen);
+            result = Value::fromI32(0);
+            LOG_INSTRUCTION("ArrayLen", arrayref, RESULT(result), "Exception");
+            return { };
+        }
+
+        Location arrayLocation = loadIfNecessary(arrayref);
+        emitThrowOnNullReference(ExceptionType::NullArrayLen, arrayLocation);
+
+        result = topValue(TypeKind::I32);
+        Location resultLocation = allocateWithHint(result, arrayLocation);
+        m_jit.load32(MacroAssembler::Address(arrayLocation.asGPR(), JSWebAssemblyArray::offsetOfSize()), resultLocation.asGPR());
+
+        LOG_INSTRUCTION("ArrayLen", arrayref, RESULT(result));
+        return { };
+    }
+
+    void emitStructSet(Value structValue, const StructType& structType, uint32_t fieldIndex, Value value)
```

Comment:
I wonder if we should pass `GPRReg structValueGPR` here since it is confusing that we consume value while we do not consume structValue.